File: apps/web/src/screens/exec/exec-agg.ts
/*
 * Pure aggregation + transform logic for the Executive Dashboard screen
 * (group-C Wave-2b, gate G3).
 *
 * The prototype (pototype/exec-audit.jsx ExecDashboard) is a MOCK: the per-project
 * `roll` object (budget/actual/progress/sold/health literals) and the window.PROJECTS
 * / window.ACCEPT_ITEMS globals are all hardcoded. Per section-0 rule 3 + C10 (PLAN.md
 * Appendix C) NONE of those mock numbers are reproduced. This module instead parses
 * the OPAQUE Entity JSON returned by the real GET /analytics/portfolio handler
 * (apps/api/src/routes/analytics.ts, B-101) into typed rows, plus the pure formatters /
 * health-tone map / over-spend rule / KPI derivations the view consumes. No React, no
 * i18n, no fetch here — every lookup stays unit-testable (G3).
 *
 * WIRE REALITY (honest, never fabricated): the handler returns budget/actual from
 * cbs_budget, progress from the phase-progress built% source, sold% from sales_unit,
 * and health as the STORED curated label surfaced verbatim (B-102). Where a source is
 * genuinely absent (no curated health / no sales units) the handler returns an HONEST
 * null; the parsers keep that null so the view can render an em-dash, never a value.
 * The accept-center strip has NO endpoint at all -> the view renders honest em-dashes.
 *
 * No raw Thai byte lives in this source (B-073): the two stored health labels this
 * module must compare against are written as \u escapes (mirrors dashboard-agg BAHT).
 */

/** An opaque contract Entity — the /analytics/portfolio body is `{ [k]: unknown }`. */
export type Ent = Record<string, unknown>;

/* ── Opaque-Entity field readers (defensive — the contract is Entity/opaque) ──── */

/** Finite number at `key`, else 0. */
export function entNum(e: Ent | undefined, key: string): number {
  const v = e?.[key];
  const n = typeof v === "number" ? v : Number(v);
  return Number.isFinite(n) ? n : 0;
}

/** Finite number at `key`, else null (distinguishes an honest gap from a real 0). */
export function entNumOrNull(e: Ent | undefined, key: string): number | null {
  const v = e?.[key];
  if (v == null) return null;
  const n = typeof v === "number" ? v : Number(v);
  return Number.isFinite(n) ? n : null;
}

/** Non-empty string at `key`, else null. */
export function entStr(e: Ent | undefined, key: string): string | null {
  const v = e?.[key];
  return typeof v === "string" && v.length > 0 ? v : null;
}

/** Array of Entities at `key`, else []. */
export function entArr(e: Ent | undefined, key: string): Ent[] {
  const v = e?.[key];
  return Array.isArray(v) ? (v as Ent[]) : [];
}

/* ── Parsed row types ────────────────────────────────────────────────────────── */

/** GET /analytics/portfolio `totals` — the portfolio KPI aggregates. */
export interface PortfolioTotals {
  budgetTotal: number;
  actualTotal: number;
  avgProgress: number;
  atRiskCount: number;
  currencyCode: string | null;
}

/** One `projects[]` row (health/sold_pct are honest-null wire gaps). */
export interface PortfolioProject {
  projectId: string;
  name: string | null;
  /** realestate | solar | civil | service — null when the project has no type. */
  typeKey: string | null;
  budget: number;
  actual: number;
  progressPct: number;
  /** Stored curated label surfaced verbatim (B-102), or null when uncurated. */
  health: string | null;
  soldPct: number | null;
}

/** One `type_mix[]` row — the budget sum grouped by project type. */
export interface TypeMixRow {
  typeKey: string;
  budgetSum: number;
}

/** The whole GET /analytics/portfolio rollup. */
export interface Portfolio {
  totals: PortfolioTotals;
  projects: PortfolioProject[];
  typeMix: TypeMixRow[];
}

/* ── Parser (opaque Entity → typed) ───────────────────────────────────────────── */

export function parsePortfolio(e: Ent | undefined): Portfolio | null {
  if (!e) return null;
  const totalsEnt = (e.totals as Ent | undefined) ?? undefined;
  return {
    totals: {
      budgetTotal: entNum(totalsEnt, "budget_total"),
      actualTotal: entNum(totalsEnt, "actual_total"),
      avgProgress: entNum(totalsEnt, "avg_progress"),
      atRiskCount: entNum(totalsEnt, "at_risk_count"),
      currencyCode: entStr(totalsEnt, "currency_code"),
    },
    projects: entArr(e, "projects").map((p) => ({
      projectId: entStr(p, "project_id") ?? "",
      name: entStr(p, "name"),
      typeKey: entStr(p, "type_key"),
      budget: entNum(p, "budget"),
      actual: entNum(p, "actual"),
      progressPct: entNum(p, "progress_pct"),
      health: entStr(p, "health"),
      soldPct: entNumOrNull(p, "sold_pct"),
    })),
    typeMix: entArr(e, "type_mix").map((m) => ({
      typeKey: entStr(m, "type_key") ?? "",
      budgetSum: entNum(m, "budget_sum"),
    })),
  };
}

/* ── Formatters ───────────────────────────────────────────────────────────────── */

/**
 * The prototype money formatter fmt (pototype/ds.jsx:4-5):
 * `Intl.NumberFormat("th-TH", { maximumFractionDigits: 0 })`. The exec KPIs / table
 * cells call fmt() over MILLIONS values (its `roll` budgets are e.g. `248`, one unit =
 * one million baht), so the wire figure (THB) is scaled by 1e6 first, mirroring the
 * mock's unit exactly.
 */
const MILLION_FMT = new Intl.NumberFormat("th-TH", { maximumFractionDigits: 0 });

/** A THB amount → the prototype's millions display (fmt over value ÷ 1e6). */
export function fmtMillions(baht: number): string {
  if (!Number.isFinite(baht)) return "0";
  return MILLION_FMT.format(baht / 1e6);
}

/* ── Health tone (pototype/exec-audit.jsx:28 healthTone) ──────────────────────────
 * The stored health label is SERVER DATA (Thai) surfaced verbatim; these two
 * constants are the byte-exact good / watch labels written as \u escapes so no raw
 * Thai byte lands in source (B-073, mirrors dashboard-agg BAHT escape):
 *   HEALTH_OK    -> the "good" label     (U+0E14 U+0E35)
 *   HEALTH_WATCH -> the "watch" label    (U+0E40 U+0E1D U+0E49 U+0E32 U+0E23 U+0E30 U+0E27 U+0E31 U+0E07)
 */
const HEALTH_OK = "\u0E14\u0E35";
const HEALTH_WATCH = "\u0E40\u0E1D\u0E49\u0E32\u0E23\u0E30\u0E27\u0E31\u0E07";

/**
 * Health chip colour by stored label (exec-audit.jsx:28: good→ok · watch→warn ·
 * else→danger). A null (uncurated) health is NOT a value — it returns the neutral
 * tone, and the view renders an honest em-dash instead of a chip (section-0 rule 3).
 */
export function healthTone(health: string | null): string {
  if (health == null) return "var(--text-3)";
  if (health === HEALTH_OK) return "var(--ok)";
  if (health === HEALTH_WATCH) return "var(--warn)";
  return "var(--danger)";
}

/* ── Presentational rules + KPI derivations ───────────────────────────────────── */

/**
 * The "over" spend colouring rule (exec-audit.jsx:89): actual > budget × 0.9 → the
 * spend cell renders in danger. Presentational only (B-102: this 0.9 threshold is NOT
 * the health rule — health is the stored label; this only colours the cell).
 */
export function overSpend(actual: number, budget: number): boolean {
  return actual > budget * 0.9;
}

/**
 * Whole-percent actual-of-budget (exec-audit.jsx:41 `Math.round(totActual/totBudget*100)`)
 * — the actual-total KPI delta + its "% of budget" sub. 0-guarded (empty portfolio → 0).
 */
export function actualPctOfBudget(actual: number, budget: number): number {
  if (!Number.isFinite(actual) || !Number.isFinite(budget) || budget === 0) return 0;
  return Math.round((100 * actual) / budget);
}

/**
 * Distinct project-type count across the portfolio (the total-projects KPI "n types"
 * sub). The mock uses Object.keys(PROJECT_TYPES).length (its global type catalog);
 * that catalog is not in the portfolio response, so the honest analog derivable from
 * the wire is the number of distinct non-null type_keys present.
 */
export function distinctTypeCount(projects: PortfolioProject[]): number {
  const set = new Set<string>();
  for (const p of projects) if (p.typeKey != null) set.add(p.typeKey);
  return set.size;
}

/**
 * Bar fill fraction (pototype/ds.jsx Bar:169) — `min(100, value/max*100)`, 0 when
 * max ≤ 0. Drives every progress / type-mix bar width (the type-mix "bar math":
 * a type's budget_sum as a share of the portfolio budget_total).
 */
export function barPct(value: number, max: number): number {
  if (!Number.isFinite(value) || !Number.isFinite(max) || max <= 0) return 0;
  return Math.min(100, (value / max) * 100);
}

File: apps/mobile/lib/screens/exec/exec_agg.dart
// Pure parse + honest derivations for the mobile executive dashboard (route
// `exec`, pototype/mobile-screens.jsx MExecDashboard L647-716). money = NONE in
// the sense that this screen ORIGINATES nothing: it displays money, and every
// figure it displays is a SERVER value read verbatim. Nothing here sums, averages,
// accumulates or otherwise derives a monetary total.
//
// No Flutter, no Dio, no i18n imports — every derivation stays unit-testable. The
// one import is the merged approvals-inbox agg, because this screen's approvals
// section reads the SAME `GET /dashboard/approvals-inbox` payload as the merged
// `inbox` screen: it reuses that module's row shape (InboxRow / parseInbox /
// formatMoney) rather than inventing a second reading of one payload.
//
// ---------------------------------------------------------------------------
// WHAT THE PROTOTYPE SHOWS vs WHAT THE WIRE CARRIES (PLAN.md §0 rule 3 + rule 4)
// ---------------------------------------------------------------------------
// The prototype is a MOCK end to end — every number on it is a literal in the
// source. Three feeds are real, and they cover exactly two of the four blocks:
//
//   GET /dashboard/summary            (apps/api/src/routes/dashboard.ts L860)
//   GET /dashboard/approvals-inbox    (dashboard.ts L862)
//   GET /boq/reports/evm              (apps/api/src/routes/boq-reports.ts L519)
//
// REAL (drawn from the wire):
//   * the S-curve            → /boq/reports/evm `series[]` (pv + ev per period).
//   * the approvals section  → /dashboard/approvals-inbox rows + their count.
//
// HONEST em-dash (no feed exists — NEVER fabricated, see BLOCKERS.md B-299):
//   * the hero's cumulative-sales figure, its units-sold ratio and its YoY delta
//     (L657-661). No endpoint serves a sales total, a unit count, or a
//     year-over-year delta. The `/sales/*` routes are per-row LIST endpoints
//     (land-sales.ts L925-930: leads / loans / bookings / contracts / downs), so
//     the only way to produce "284.5 M" from them would be to SUM booking rows in
//     the client — precisely the money-origination the iron rules forbid.
//   * all four KPI tiles (L666-669) — see kpi notes on ExecKpis below.
//   * the S-curve legend percentages "plan 65% / actual 62%" (L690-691). Percent
//     complete in EVM is EV/BAC, and BAC is a REAL stored column
//     (packages/db/src/schema/finance.ts L922, "the project total") that
//     loadEvmSeries already loads (evm-series.ts L49) — but the /boq/reports/evm
//     handler does NOT put it in its response body (boq-reports.ts evm returns
//     series/spi/cpi/currency_code only). Dividing by anything else — the
//     summary's budget_total, the series max — would be inventing a denominator,
//     so both percentages are em-dashed and B-299 files the one-field gap.
//
// The prototype's `MExec` header eyebrow is a person's name (L650) and its bell
// carries an unread DOT (L653); neither has a wire (AppServices exposes a
// bearer token, not a user profile; no unread-count feed is read here), so the
// eyebrow is omitted (the merged sales-crm precedent) and the dot is dropped — a
// dot IS the claim "you have unread notifications", and a claim cannot be
// em-dashed the way a missing value can.
import '../approvals_inbox/approvals_inbox_agg.dart';

/// An opaque contract Entity / response object — every dashboard + report payload
/// is modelled as the opaque `Entity` in openapi.yaml (no declared fields), so it
/// is read as a raw map exactly as the merged screens read theirs.
typedef ExecEnt = Map<String, Object?>;

/// Non-empty string at the first present key, else null (never "" — the view
/// em-dashes a null, and a blank is the same absence).
String? execStr(ExecEnt e, List<String> keys) {
  for (final String key in keys) {
    final Object? v = e[key];
    if (v is String && v.isNotEmpty) return v;
  }
  return null;
}

/// Finite number at the first present key, else null (absent / non-numeric /
/// NaN / infinite are all "unreadable" → the caller em-dashes or drops).
double? execNum(ExecEnt e, List<String> keys) {
  for (final String key in keys) {
    final Object? v = e[key];
    if (v is num && v.isFinite) return v.toDouble();
    if (v is String && v.isNotEmpty) {
      final double? parsed = double.tryParse(v);
      if (parsed != null && parsed.isFinite) return parsed;
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// GET /dashboard/summary
// ---------------------------------------------------------------------------

/// The ONE field this screen reads from `GET /dashboard/summary`: the id of the
/// project the dashboard is about (the caller's PRIMARY project when no
/// ?project_id is passed — dashboard.ts summary → resolvePrimaryProject).
///
/// WHY ONLY THIS FIELD. The summary payload is a BUDGET KPI set —
/// budget_total / actual_total / committed_total / remaining_total / health_score
/// plus project meta (name, type, active phase, status, as_of). The exec
/// prototype displays no budget figure and no project meta: its hero is a SALES
/// total and its tiles are cash-flow / pending-docs / deadlines / monthly sales.
/// Rendering budget_total under the hero's "cumulative sales" label would be
/// relabelling one quantity as another — a fabrication, not a port — so no
/// summary money reaches the screen.
///
/// It is still a load-bearing read, and not an ornamental one: `GET
/// /boq/reports/evm` WITHOUT a project_id returns every owned project's snapshots
/// merged into one list (evm-series.ts loadEvmSeries: `projectId ? eq(...) :
/// undefined`), which would interleave two projects' PV/EV under the same period
/// key and draw a curve that belongs to no project. Scoping the EVM read to this
/// id is what makes the S-curve a real single-project curve.
///
/// Null when the payload is unreadable or the tenant has no project (summary
/// returns `project_id: null` in that case) → the caller falls back to the
/// tenant-wide read and the S-curve renders whatever honestly exists.
String? parseSummaryProjectId(ExecEnt? body) {
  if (body == null) return null;
  return execStr(body, <String>['project_id', 'projectId']);
}

// ---------------------------------------------------------------------------
// GET /boq/reports/evm — the S-curve
// ---------------------------------------------------------------------------

/// One EVM period as the server reports it. Both values are SERVER money in the
/// series' own currency, read verbatim — the schema documents pv and ev as
/// CUMULATIVE to the period end (finance.ts L913-916), which is exactly what an
/// S-curve plots, so nothing is accumulated, interpolated or scaled here.
class EvmPoint {
  const EvmPoint({
    required this.period,
    required this.monthOrdinal,
    required this.pv,
    required this.ev,
  });

  /// The server's own `period_label` ('YYYY-MM'), rendered verbatim on the axis.
  final String period;

  /// The period's real position in time as a month count (year*12 + month-1).
  /// This is what places a point on the x axis, so a MISSING month leaves a real
  /// gap in the curve instead of being silently closed up.
  final int monthOrdinal;

  /// Planned Value, cumulative (the prototype's dashed "plan" curve).
  final double pv;

  /// Earned Value, cumulative (the prototype's solid "actual" curve). EV — not
  /// AC — is the progress counterpart of PV; AC is a COST series and the
  /// prototype draws only two curves, so drawing a third would be a redesign.
  final double ev;
}

/// Parse a 'YYYY-MM' period key into a month ordinal, or null when the label is
/// not that shape. Nothing is guessed: an unparseable label cannot be placed on a
/// time axis at all.
int? monthOrdinal(String period) {
  if (period.length != 7 || period[4] != '-') return null;
  final int? year = int.tryParse(period.substring(0, 4));
  final int? month = int.tryParse(period.substring(5, 7));
  if (year == null || month == null) return null;
  if (month < 1 || month > 12) return null;
  return year * 12 + (month - 1);
}

/// Parse the `series` array of `GET /boq/reports/evm` into plottable points.
///
/// A row is KEPT only when all three of its plotting facts are readable: a
/// 'YYYY-MM' period_label, a finite pv and a finite ev. A row missing any of them
/// is DROPPED rather than defaulted, because every fallback would be an invention
/// — a 0 would draw a real point on the curve that the series does not contain,
/// and an index-based x would place the point at a time it does not have. (The
/// server makes this a defensive branch only: pv/ev are NOT NULL numerics with a
/// 0 default and period is written as a 'YYYY-MM' key.)
///
/// Server order is preserved (loadEvmSeries sorts by period ASC, and for
/// 'YYYY-MM' keys a lexical sort IS chronological).
List<EvmPoint> parseEvmSeries(Object? series) {
  if (series is! List) return const <EvmPoint>[];
  final List<EvmPoint> out = <EvmPoint>[];
  for (final Object? row in series) {
    if (row is! Map) continue;
    final ExecEnt e = row.map<String, Object?>(
      (Object? k, Object? v) => MapEntry<String, Object?>('$k', v),
    );
    final String? period = execStr(e, <String>['period_label', 'periodLabel']);
    if (period == null) continue;
    final int? ord = monthOrdinal(period);
    if (ord == null) continue;
    final double? pv = execNum(e, <String>['pv']);
    final double? ev = execNum(e, <String>['ev']);
    if (pv == null || ev == null) continue;
    out.add(EvmPoint(period: period, monthOrdinal: ord, pv: pv, ev: ev));
  }
  return out;
}

/// A normalised chart coordinate. Plain data (no dart:ui here — this module stays
/// Flutter-free); [x] and [y] are both 0..1 fractions the painter maps to pixels.
class ChartXY {
  const ChartXY(this.x, this.y);

  /// 0 = the earliest plotted period, 1 = the latest, positioned by REAL elapsed
  /// months so a gap in the series is a gap on the axis.
  final double x;

  /// 0 = the value baseline, 1 = the largest value in the series. The painter
  /// flips it (screen y grows downward).
  final double y;
}

/// The S-curve's geometry, derived from the real series only.
///
/// Both curves carry exactly one vertex per SERVED point — no smoothing control
/// points, no resampling, no zero-filled periods. The painter joins the vertices
/// with straight segments precisely because a bezier (which the prototype uses,
/// L682-683) would bow the line through values the series never reported.
class EvmChartGeometry {
  const EvmChartGeometry({
    required this.plan,
    required this.actual,
    required this.todayX,
    required this.firstLabel,
    required this.lastLabel,
  });

  /// The PV vertices (prototype's dashed curve), in period order.
  final List<ChartXY> plan;

  /// The EV vertices (prototype's solid curve), in period order. Always the same
  /// length as [plan] — both come from the same kept rows.
  final List<ChartXY> actual;

  /// x of the "today" marker, or NULL when the current month falls outside the
  /// plotted range. The prototype pins this line at a fixed x (L684); here it is
  /// only ever drawn at the real position of the real current month, and when
  /// today is off the chart it is not drawn at all.
  final double? todayX;

  /// The first / last plotted period labels — the axis ends. Null when empty.
  final String? firstLabel;
  final String? lastLabel;

  /// Nothing plottable — the view renders the honest empty state, never an axis
  /// with no curve on it.
  bool get isEmpty => plan.isEmpty;

  /// A single period: there is no line to draw between two moments, so the
  /// painter marks the one real point and draws no segment (a segment would
  /// imply a second value).
  bool get isSinglePoint => plan.length == 1;
}

/// An empty geometry (no series, or nothing plottable in it).
const EvmChartGeometry kEmptyEvmGeometry = EvmChartGeometry(
  plan: <ChartXY>[],
  actual: <ChartXY>[],
  todayX: null,
  firstLabel: null,
  lastLabel: null,
);

/// Project [points] onto the unit square, and locate "today" as of [now].
///
/// x — from the real month ordinals, so the spacing between two points is the
/// real number of months between them (a skipped period leaves a wider segment,
/// never a fabricated intermediate point). A single point sits at the centre.
///
/// y — anchored at the value baseline `min(0, lowest)` and topped at
/// `max(0, highest)` across BOTH series, so a non-negative series is drawn
/// against a true zero and its shape is not exaggerated by a floating floor. A
/// completely flat series (range 0) is drawn flat on the baseline rather than
/// dividing by zero.
EvmChartGeometry evmChartGeometry(List<EvmPoint> points, DateTime now) {
  if (points.isEmpty) return kEmptyEvmGeometry;

  int minOrd = points.first.monthOrdinal;
  int maxOrd = points.first.monthOrdinal;
  double lowest = 0;
  double highest = 0;
  for (final EvmPoint p in points) {
    if (p.monthOrdinal < minOrd) minOrd = p.monthOrdinal;
    if (p.monthOrdinal > maxOrd) maxOrd = p.monthOrdinal;
    if (p.pv < lowest) lowest = p.pv;
    if (p.ev < lowest) lowest = p.ev;
    if (p.pv > highest) highest = p.pv;
    if (p.ev > highest) highest = p.ev;
  }

  final int span = maxOrd - minOrd;
  final double valueRange = highest - lowest;

  double xOf(int ord) => span == 0 ? 0.5 : (ord - minOrd) / span;
  double yOf(double v) => valueRange == 0 ? 0 : (v - lowest) / valueRange;

  final List<ChartXY> plan = <ChartXY>[];
  final List<ChartXY> actual = <ChartXY>[];
  for (final EvmPoint p in points) {
    final double x = xOf(p.monthOrdinal);
    plan.add(ChartXY(x, yOf(p.pv)));
    actual.add(ChartXY(x, yOf(p.ev)));
  }

  // The current month as an ordinal on the same axis. Drawn ONLY when it lies
  // inside the plotted range — a marker beyond the last point would assert the
  // axis extends into a period the series does not cover.
  final int todayOrd = now.year * 12 + (now.month - 1);
  final double? todayX = (todayOrd < minOrd || todayOrd > maxOrd)
      ? null
      : xOf(todayOrd);

  return EvmChartGeometry(
    plan: plan,
    actual: actual,
    todayX: todayX,
    firstLabel: points.first.period,
    lastLabel: points.last.period,
  );
}

// ---------------------------------------------------------------------------
// The KPI tiles + hero
// ---------------------------------------------------------------------------

/// The four KPI tiles (L666-669) and the hero figures (L657-661).
///
/// EVERY slot is honestly unknown, and each for its own verified reason. This
/// type exists so the gap is DECLARED in one place and the view cannot quietly
/// grow a fabricated value later:
///
///   * `cashFlow7d` (L666, "cash flow 7 days") — a 7-day cash forecast IS served,
///     by `GET /dashboard/cashflow-forecast` (dashboard.ts L865). That endpoint
///     is outside this slice's three feeds, so nothing is read for it and the
///     tile is em-dashed rather than guessed. This is the one tile that is a
///     wiring gap rather than a backend gap (B-299).
///   * `pendingDocs` (L667, "documents awaiting approval") — the only pending
///     count available is `GET /dashboard/approvals-inbox`, and that list is
///     TIER-FILTERED to what the CALLER may act on (dashboard.ts approvalsInbox →
///     callerApprovalLevel / canApprove). Its count is therefore "awaiting MY
///     approval", not "awaiting approval"; the prototype itself treats them as
///     different quantities (tile 17 vs section 3). Putting the caller's subset
///     under the unscoped label would understate the true backlog, so the tile is
///     em-dashed and the REAL count is shown where it belongs — on the section
///     that actually says "awaiting my approval" ([approvalsCount]).
///   * `projectsDue` (L668, "projects near deadline") — no endpoint returns a
///     due-soon project count.
///   * `salesMonth` (L669, "sales this month") — no sales-total endpoint exists;
///     see the module header on why summing `/sales/*` rows is forbidden.
///   * `salesCumulative` / `unitsSold` / `unitsTotal` / `yoyPercent` — the hero
///     (L657-661); same missing sales + unit feeds.
///
/// Every field is `null` today and the view renders an em-dash for each. They are
/// modelled rather than hard-coded as dashes so that the day a feed lands, the
/// wiring point is obvious and typed.
class ExecKpis {
  const ExecKpis({
    this.salesCumulative,
    this.unitsSold,
    this.unitsTotal,
    this.yoyPercent,
    this.cashFlow7d,
    this.pendingDocs,
    this.projectsDue,
    this.salesMonth,
  });

  /// Hero: cumulative sales for the year (SERVER value when one ever exists).
  final double? salesCumulative;

  /// Hero: units sold / total units.
  final int? unitsSold;
  final int? unitsTotal;

  /// Hero: year-over-year delta, in percent.
  final double? yoyPercent;

  /// Tile 1: net 7-day cash flow (SERVER value).
  final double? cashFlow7d;

  /// Tile 2: count of documents awaiting approval, tenant-wide.
  final int? pendingDocs;

  /// Tile 3: count of projects near their deadline.
  final int? projectsDue;

  /// Tile 4: sales booked this month (SERVER value).
  final double? salesMonth;
}

/// The whole screen's honest state.
class ExecDashboard {
  const ExecDashboard({
    required this.kpis,
    required this.chart,
    required this.approvals,
  });

  /// Hero + tile figures (all honestly unknown today — see [ExecKpis]).
  final ExecKpis kpis;

  /// The S-curve geometry, from the REAL EVM series.
  final EvmChartGeometry chart;

  /// The caller's pending-and-actionable docs, server order (newest first),
  /// parsed by the merged inbox agg — same payload, one reading.
  final List<InboxRow> approvals;

  /// The approvals section's count. This is the list length, which is exactly the
  /// envelope's `total`: listEnvelope sets `total = rows.length` for these
  /// single-page handlers (apps/api/src/routes/list-envelope.ts L32-38), so
  /// reading the length is not an approximation of the server's count — it IS it.
  int get approvalsCount => approvals.length;
}

/// Assemble the screen state from the payloads the three feeds return. Pure: no
/// clock of its own ([now] is injected so the "today" marker is testable and
/// never ambient).
///
/// There is no `summary` parameter on purpose. The summary payload's ONE job on
/// this screen is to yield the project id that SCOPES the EVM read
/// ([parseSummaryProjectId]), which happens in the repository before the series
/// is fetched; none of its own fields reach the view (see
/// [parseSummaryProjectId] for why its budget figures must not).
ExecDashboard buildExecDashboard({
  required Object? evmSeries,
  required List<InboxEnt> approvalRows,
  required DateTime now,
}) {
  final List<InboxRow> approvals = parseInbox(approvalRows);
  return ExecDashboard(
    kpis: const ExecKpis(),
    chart: evmChartGeometry(parseEvmSeries(evmSeries), now),
    approvals: approvals,
  );
}
